### PR TITLE
rt: bring back a public Handle type

### LINKED ISSUE #2965: Replacement for Handle::current
With the removal of `Handle`, there is no equivalent to `Handle::current`. For example, the following is no longer easily possible:
```rust
let handle = Handle::current();
std::thread::spawn(move || {
    handle.block_on(...);
});
```

### PR DESCRIPTION
**Pull request by Keruspe**
- make the `Handle` type public
- make `Runtime::spawn` and `Runtime::spawn_blocking` just thin wrappers around new public methods from `Handle`
- introduce `handleEnterGuard` to bring back `Handle::enter` (not sure if it's the right call here)
- didn't bring back `Handle::block_on` (not sure if that's something we want, feels less trivial to do, and maybe not that useful)
- brought back `Handle::current` and `Handle::try_current`

Fixes #2965

### CODE REVIEW

**Comment by Darksonn** [comment]
Please prefer to push more commits rather than force-pushing as it makes it very difficult for me to see what has changed since last time I looked.  
Do not worry about the resulting history on the master branch, as your PR will be squashed on merge.

**Comment by Keruspe** [comment]
Those were just to fix a formatting issue

**Comment by Keruspe** [comment]
@Darksonn to be fair I did not expect someone to look at it before CI was green, thus the force-pushes wrt formatting to make CI green (+ some git messup on my side which resulted in way more pushes that needed be)

**Comment by Darksonn** [comment]
Well I had not looked at it yet, but I mention it now because if I don't, experience shows that I have to later when I _do_ come back to see the changes.

**Comment by Keruspe** [comment]
Should I rebase or should I merge?

**Comment by Darksonn** [comment]
A merge commit works perfectly fine.

**Comment by Keruspe** [comment]
@hawkw you might want to look at this merge commit to make sure I did the right thing wrt your recent tracing changes

File: tokio/src/runtime/handle.rs [hunk 1]
```diff
@@ -28,21 +32,131 @@ pub(crate) struct Handle {
     pub(super) blocking_spawner: blocking::Spawner,
 }
 
+#[derive(Debug)]
+pub struct HandleEnterGuard(EnterGuard);
```
Comment (carllerche) [inline tokio/src/runtime/handle.rs:36]: We shouldn't need to define a new struct. It should be possible to modify `runtime::EnterGuard` to be backed by `Handle`. `Runtime::enter` can then delegate to `Handle::enter`.

**Review by carllerche — CHANGES_REQUESTED** [review]
Thanks :+1: looking good. I left one request inline.

File: tokio/src/runtime/handle.rs [hunk 2]
```diff
@@ -28,21 +32,131 @@ pub(crate) struct Handle {
...
+    /// // Spawn a blocking function onto the runtime using the handle
+    /// handle.spawn_blocking(|| {
+    ///     println!("now running on a worker thread");
+    /// });
+    /// # }
     #[cfg_attr(tokio_track_caller, track_caller)]
-    pub(crate) fn spawn_blocking<F, R>(&self, func: F) -> JoinHandle<R>
+    pub fn spawn_blocking<F, R>(&self, func: F) -> JoinHandle<R>
     where
         F: FnOnce() -> R + Send + 'static,
```
Comment (Darksonn) [inline tokio/src/runtime/handle.rs:161]:
```suggestion
    where
        F: FnOnce() -> R + Send + 'static,
        R: Send + 'static,
```

File: tokio/src/runtime/handle.rs (hunk 1 above)
Comment (Keruspe) [reply tokio/src/runtime/handle.rs:36]: I wasn't convinced by this new type either. Your proposal sounds way better indeed!
Testing a fix

File: tokio/src/runtime/handle.rs [hunk 3]
```diff
@@ -28,23 +31,144 @@ pub(crate) struct Handle {
     pub(super) blocking_spawner: blocking::Spawner,
 }
 
+/// Runtime context guard.
+///
+/// Returned by [`Runtime::enter`] and [^Handle::enter^], the context guard exits
```
Comment (hawkw) [inline tokio/src/runtime/handle.rs:36]: are the ^s a typo, or does that mean something to RustDoc? to me, it _looks_ like this should be
```suggestion
/// Returned by [`Runtime::enter`] and [`Handle::enter`], the context guard exits
```

Comment (Keruspe) [reply tokio/src/runtime/handle.rs]: Definitely a typo

**Comment by Darksonn** [comment]
The changes look good to me, although I notice that `block_on` is missing.

**Comment by Keruspe** [comment]
I left block_on on purpose, since I'm not sure it's the most useful thing to have on Hande (current + spawn being the more important imo), and it would require some less trivial work to do since Runtime now uses things Handle doesn't have access to for its block_on

**Comment by Darksonn** [comment]
Well, I've probably used `Handle::block_on` more often than `Handle::spawn`. But in any case, I'm ok with leaving it for a follow-up PR.

**Review by carllerche — APPROVED** [review]
thanks

**Comment by stuhood** [comment]
> didn't bring back Handle::block_on (not sure if that's something we want, feels less trivial to do, and maybe not that useful)

`block_on` was, AFAICT, the primary motivating case for bringing back `Handle` in #2965.

Without `Handle::block_on`, I think that there is no way for a `tokio::main` environment to ever call `block_on`, because there is no way to get access to the current `Runtime`. Preferable to having the `Handle`/`Runtime` split in the public API though would be a `Runtime::current` method.

Our use of `Runtime`s is generic on how the `Runtime` is constructed, in that in tests we use the `tokio::test` macro to create the runtime (and thus don't have a reference to it to ever call `block_on` with), while in production we use the `Builder` to construct a `Runtime` (and thus have a reference to use `block_on` with). We used to resolve this by using `Handle` as the lowest common denominator ("regardless of how your `Runtime` was created, a `Handle` is always available one way or another").

**Comment by Keruspe** [comment]
@stuhood I still think we should leave block_on for a followup pull request as it can get way more complicated to implement than all of this
I'm not saying it sholdn't be done, I just think that the current state already solves at least some use cases without involving too much complexity.

To be fair, I have a nearly finished port of block_on locally, but in itself it involves bigger changes than this entire pull request. The last blocker left are these two lines that are non trivial to port since the BasicScheduler is parametric against Park: https://github.com/tokio-rs/tokio/blob/20a2b9e26316e704d16c62a28dfb74a505b92b25/tokio/src/runtime/basic_scheduler.rs#L144-L145

**Review by Darksonn — APPROVED** [review]

**Pull request merged**